yoyodyne/models/modules/lstm.py

```diff
@@ -223,3 +223,167 @@ def forward(
+        ).squeeze(-1)
+        # Gets probability of alignments.
+        alignment_probs = nn.functional.softmax(alignment_scores, dim=-1)
+        # Mask padding.
+        alignment_probs = alignment_probs * (~encoder_mask) + 1e-7
+        alignment_probs = alignment_probs / alignment_probs.sum(
+            dim=-1, keepdim=True
+        )
+        # Expands over all time steps. Log probs for quicker computation.
+        return (
+            alignment_probs.log()
+            .unsqueeze(1)
+            .expand(-1, encoder_out.shape[1], -1)
+        )
+
+    def forward(
+        self,
+        symbol: torch.Tensor,
+        last_hiddens: torch.Tensor,
+        encoder_out: torch.Tensor,
+        encoder_mask: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Single decode pass.
+
+        Args:
+            symbol (torch.Tensor): previously decoded symbol of shape (B x 1).
+            last_hiddens (Tuple[torch.Tensor, torch.Tensor]): last hidden
+                states from the decoder of shape
+                (1 x B x decoder_dim, 1 x B x decoder_dim).
+            encoder_out (torch.Tensor): encoded input sequence of shape
+                (B x seq_len x encoder_dim).
+            encoder_mask (torch.Tensor): mask for the encoded input batch of
+                shape (B x seq_len).
+
+        Returns:
+            output (base.ModuleOutput): Dataclass containing step-wise
+                emission probs, alignment matrix, and hidden states of decoder.
+        """
+        # Encodes current symbol.
+        embedded = self.embed(symbol)
+        decoded, hiddens = self.module(embedded, last_hiddens)
+        # Gets emission probabilities over each hidden state (source symbol).
+        output = decoded.expand(-1, encoder_out.shape[1], -1)
+        output = torch.cat([output, encoder_out], dim=-1)
+        output = self.output_proj(output)
+        # Gets transition probabilities (alignment) for current states.
+        alignment = self._alignment_step(decoded, encoder_out, encoder_mask)
+        return base.ModuleOutput(
+            output=output, embeddings=alignment, hiddens=hiddens
+        )
+
+    def get_module(self) -> nn.LSTM:
+        return nn.LSTM(
+            self.embedding_size,
+            self.hidden_size,
+            num_layers=self.layers,
+            dropout=self.dropout,
+            batch_first=True,
+            bidirectional=self.bidirectional,
+        )
+
+    @property
+    def output_size(self) -> int:
+        return self.decoder_input_size + self.hidden_size
+
+    @property
+    def name(self) -> str:
+        return "HMM LSTM"
+
+
+class ContextHMMLSTMDecoder(HMMLSTMDecoder):
+    def __init__(self, *args, attention_context, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.delta = attention_context
+        self.alignment_proj = nn.Linear(
+            self.hidden_size * 2, (attention_context * 2) + 1
```

Comment:
Does this mean this implements strictly a 1st order HMM? Could be good to note that.